### ISSUE TITLE
How can we help users change number type more easily?

### ISSUE THREAD

**Issue by glwagner** [CliMA/Oceananigans.jl#3800]
Running a simulation with `FT=Float32` is fairly onerous because the number type has to be passed to a huge number of model components individually. Eg we need to write

```julia
using Oceananigans
using SeawaterPolynomials: TEOS10EquationOfState

FT = Float32
grid = RectilinearGrid(FT, size=(1, 1, 1), extent=(1, 1, 1))
coriolis = FPlane(FT, f=1)
equation_of_state = TEOS10EquationOfState(FT)
buoyancy = SeawaterBuoyancy(FT; equation_of_state)
```

Here are two ideas to make this easier:

## Use an environment variable

For this we introduce an environment variable, say `OCEANANIGANS_NUMBER_TYPE`. Then we can write a small utility:

```julia
function global_number_type(default=Float64)
    user_number_type = get("OCEANANIGANS_NUMBER_TYPE", ENV, nothing)

    if isnothing(user_number_type)
        FT = default
    else
        try
            FT = eval(Symbol(user_number_type))
        catch
            error("The environment variable OCEANANIGANS_NUMBER_TYPE=$user_number_type is not supported!")
        end
    end

    return FT
end
```

This also permits a script-based approach because users can write

```julia
ENV["OCEANANIGANS_NUMBER_TYPE"] = "Float32"

grid = RectilinearGrid(size=(1, 1, 1), extent=(1, 1, 1))
coriolis = FPlane(f=1)
equation_of_state = TEOS10EquationOfState()
buoyancy = SeawaterBuoyancy(; equation_of_state)
```

## Introducing an internal "settings" dictionary

Then users might write

```julia
Oceananigans.settings[:default_number_type] = Float32
```

This avoids the string-based restriction of environment variables, so may generalize to more types of settings should we need that. It's a slippery slope of course, we want to keep this to a minimum because global settings inhibit programmability (ie one cannot have multiple simulations in a single instance with different settings). Here, the setting only controls a default, so it remains possible to have two simulations in one instance with different number types --- they just cannot be set up simultaneously (or one has to do more work to set the number type "manually").

External packages can opt into using this only if they depend on Oceananigans. Which may be ok, in fact. External packages that extend Oceananigans of course must depend on Oceananigans. There is just the question of what to do with SeawaterPolynomials, where the dependency goes the other direction...

I thought of the environment-variable approach so that SeawaterPolynomials could implement it without depending on Oceananigans. But as I think about it more, it is hacky for SeawaterPolynomials to contain an environment variable called OCEANANIGANS_NUMBER_TYPE. So now I think the second approach may be better in fact, and to support this kind of thing with SeawaterPolynomials, perhaps what we need is a way to extend the key SeawaterPolynomials constructors inside Oceananigans somehow.

I also think it would be nice to have `validate_number_type` utility for comparing number types to the grid eltype to catch inconsistent number types.

@ali-ramadhan may be interested since it I noticed some Float32's creeping into scripts...

**Comment by Yixiao-Zhang** [comment]
Thank you for bringing this up! I am interested in running simulations with `FT=Float32`. In many of my simulations, the spatial resolution is limited by GPU memory, and thus using `FT=Float32` will hopefully double the number of grid points for a given GPU. \(Although multiple-GPU should be the solution.\)

It seems to me that the document does not tell people how to setup a simulation with `FT=Float32`. It only tells people to avoid mixed-precision operations. I am willing to give a try if there is a detailed instruction.

**Comment by glwagner** [comment]
Which document are you referring to @Yixiao-Zhang ? 

I think its important to enable Float32 for science applications. It's important enough that I think this could be the subject of a paper, especially since it does not seem trivially achievable.

**Comment by Yixiao-Zhang** [comment]
I am referring to [the legacy model setup section](https://clima.github.io/OceananigansDocumentation/stable/model_setup/number_type/).

**Comment by glwagner** [comment]
Heh, I agree that says nothing!

Do you have any thoughts about a preferred user interface, among the options I give above (or perhaps you have ideas for another design)? That's the first design question to answer before writing any code. Once we settle on an interface I think the code-writing will be fairly straightforward in this case.

**Comment by Yixiao-Zhang** [comment]
As a user, I prefer an internal "settings" dictionary over an environment variable. Actually, I think it would be better to define a function `Oceananigans.set_default_number_type(FT)` for users to access the "setting" dictionary. One reason is that an error will be thrown when the user makes a typo.

Another option is to pass `FT=Float32` to the model and then let the constructor convert related components to `Float32` \(or `Float64` if `FT=Float64` is passed\). I feel it might be the most user-friendly interface. But obviously it will lots of efforts to implement.

**Comment by glwagner** [comment]
> Another option is to pass FT=Float32 to the model and then let the constructor convert related components to Float32 (or Float64 if FT=Float64 is passed). I feel it might be the most user-friendly interface. But obviously it will lots of efforts to implement.

the issue is that this cannot handle the `grid`, which has to be constructed before the model

> Actually, I think it would be better to define a function Oceananigans.set_default_number_type(FT) for users to access the "setting" dictionary. One reason is that an error will be thrown when the user makes a typo.

I think that's a nice idea. It's also maybe more sensible right now if there is only one default.

**Comment by ali-ramadhan** [comment]
I agree the current state of affairs puts a lot of responsibility on the user to ensure they pass `Float32` to every constructor that requires it, which is not very _friendly_ especially  to new users who aren't familiar with the code.

I also agree that having a way to specify the default number type in Julia is better than using an environment variable. Besides the limitation of strings, I feel like it would stray from Oceananigans.jl's script-based interface. If you share a script with someone now, you also need to inform them to run with specific environment variables. With enough environment variables you'll want a `.env` file.

Agree with @Yixiao-Zhang  that with `Oceananigans.set_default_number_type(FT)` we can do some checks and throw appropriate warnings compared with `Oceananigans.settings[:default_number_type] = Float32` (assuming it's a vanilla dictionary).

> Here, the setting only controls a default, so it remains possible to have two simulations in one instance with different number types --- they just cannot be set up simultaneously (or one has to do more work to set the number type "manually").

Glad you brought this up! Definitely a use case for an expert/power user but I can think of a few use cases (e.g. nested simulations) that would be good to support.
